hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -516,59 +520,51 @@ def fixate_shrink_passes(self, passes):
         while any_ran:
             any_ran = False
 
-            # We run remove_discarded after every step to do cleanup
+            reordering = {}
+
+            # We run remove_discarded after every pass to do cleanup
             # keeping track of whether that actually works. Either there is
             # no discarded data and it is basically free, or it reliably works
             # and deletes data, or it doesn't work. In that latter case we turn
             # it off for the rest of this loop through the passes, but will
             # try again once all of the passes have been run.
             can_discard = self.remove_discarded()
 
-            successful_passes = set()
-
             for sp in passes:
-                # We run each pass until it has failed a certain number
-                # of times, where a "failure" is any step where it made
-                # at least one call and did not result in a shrink.
-                # This gives passes which work reasonably often more of
-                # chance to run.
-                failures = 0
-                successes = 0
-
-                # The choice of 3 is fairly arbitrary and was hand tuned
-                # to some particular examples. It is very unlikely that
-                # is the best choice in general, but it's not an
-                # unreasonable choice: Making it smaller than this would
-                # give too high a chance of an otherwise very worthwhile
-                # pass getting screened out too early if it got unlucky,
-                # and making it much larger than this would result in us
-                # spending too much time on bad passes.
-                max_failures = 3
+                if can_discard:
+                    can_discard = self.remove_discarded()
 
+                before_sp = self.shrink_target
+
+                # Run the shrink pass until it fails to make any progress
+                # ten times in a row. This implicitly boosts shrink passes
+                # that are more likely to work.
+                failures = 0
+                max_failures = 20
                 while failures < max_failures:
-                    prev_calls = self.calls
                     prev = self.shrink_target
-                    if sp.step():
-                        any_ran = True
-                    else:
+                    initial_calls = self.calls
+                    if not sp.step(random=failures >= max_failures // 2):
                         break
```

Comment:
This `break` could use a quick comment explaining that we are skipping this pass because it's currently exhausted.